File: a-sync-relay/src/main/java/it/anyplace/sync/client/protocol/rp/RelayClient.java
/* 
 * Copyright (C) 2016 Davide Imbriaco
 *
 * This Java file is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package it.anyplace.sync.client.protocol.rp;

import com.google.common.base.MoreObjects;
import com.google.common.base.Preconditions;
import com.google.common.io.BaseEncoding;
import it.anyplace.sync.client.protocol.rp.beans.SessionInvitation;
import it.anyplace.sync.core.beans.DeviceAddress;
import it.anyplace.sync.core.beans.DeviceAddress.AddressType;
import it.anyplace.sync.core.configuration.ConfigurationService;
import it.anyplace.sync.core.interfaces.RelayConnection;
import it.anyplace.sync.core.security.KeystoreHandler;
import org.apache.commons.io.IOUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.ByteBuffer;

import static com.google.common.base.Preconditions.checkArgument;
import static it.anyplace.sync.core.security.KeystoreHandler.*;

/**
 *
 * @author aleph
 */
public final class RelayClient {

    private final static int MAGIC = 0x9E79BC40,
        JOIN_SESSION_REQUEST = 3,
        RESPONSE = 4,
        CONNECT_REQUEST = 5,
        SESSION_INVITATION = 6,
        RESPONSE_SUCCESS_CODE = 0;
    private final Logger logger = LoggerFactory.getLogger(getClass());
    private final KeystoreHandler keystoreHandler;

    public RelayClient(ConfigurationService configuration) {
        this.keystoreHandler = KeystoreHandler.newLoader().loadAndStore(configuration);
    }

    public RelayConnection openRelayConnection(DeviceAddress address) throws IOException, KeystoreHandler.CryptoException {
        Preconditions.checkArgument(address.getType().equals(AddressType.RELAY));
        SessionInvitation sessionInvitation = getSessionInvitation(address.getSocketAddress(), address.getDeviceId());
        return openConnectionSessionMode(sessionInvitation);
    }

    public RelayConnection openConnectionSessionMode(final SessionInvitation sessionInvitation) throws IOException {
        logger.debug("connecting to relay = {}:{} (session mode)", sessionInvitation.getAddress(), sessionInvitation.getPort());
        final Socket socket = new Socket(sessionInvitation.getAddress(), sessionInvitation.getPort());
        RelayDataInputStream in = new RelayDataInputStream(socket.getInputStream());
        RelayDataOutputStream out = new RelayDataOutputStream(socket.getOutputStream());
        {
            logger.debug("sending join session request, session key = {}", sessionInvitation.getKey());
            byte[] key = BaseEncoding.base16().decode(sessionInvitation.getKey());
            int lengthOfKey = key.length;
            out.writeHeader(JOIN_SESSION_REQUEST, 4 + lengthOfKey);
            out.writeInt(lengthOfKey);
            out.write(key);
            out.flush();
        }
        {
            logger.debug("reading relay response");
            MessageReader messageReader = in.readMessage();
            checkArgument(messageReader.getType() == RESPONSE);
            Response response = messageReader.readResponse();
            logger.debug("response = {}", response);
            checkArgument(response.getCode() == RESPONSE_SUCCESS_CODE, "response code = %s (%s) expected %s", response.getCode(), response.getMessage(), RESPONSE_SUCCESS_CODE);
            logger.debug("relay connection ready");
        }
        return new RelayConnection() {
            @Override
            public Socket getSocket() {
                return socket;
            }

            @Override
            public boolean isServerSocket() {
                return sessionInvitation.isServerSocket();
            }

        };
    }

    public SessionInvitation getSessionInvitation(InetSocketAddress relaySocketAddress, String deviceId) throws IOException, KeystoreHandler.CryptoException {
        logger.debug("connecting to relay = {} (temporary protocol mode)", relaySocketAddress);
        try (Socket socket = keystoreHandler.createSocket(relaySocketAddress, RELAY);
            RelayDataInputStream in = new RelayDataInputStream(socket.getInputStream());
            RelayDataOutputStream out = new RelayDataOutputStream(socket.getOutputStream())) {
            {
                logger.debug("sending connect request for device = {}", deviceId);
                byte[] deviceIdData = deviceIdStringToHashData(deviceId);
                int lengthOfId = deviceIdData.length;
                out.writeHeader(CONNECT_REQUEST, 4 + lengthOfId);
                out.writeInt(lengthOfId);
                out.write(deviceIdData);
                out.flush();
            }

            {
                logger.debug("receiving session invitation");
                MessageReader messageReader = in.readMessage();
                logger.debug("received message = {}", messageReader.dumpMessageForDebug());
                checkArgument(messageReader.getType() == SESSION_INVITATION, "message type mismatch, expected %s, got %s", SESSION_INVITATION, messageReader.getType());
                SessionInvitation.Builder builder = SessionInvitation.newBuilder();
                builder.setFrom(hashDataToDeviceIdString(messageReader.readLengthAndData()));
                builder.setKey(BaseEncoding.base16().encode(messageReader.readLengthAndData()));
                byte[] address = messageReader.readLengthAndData();
                if (address.length == 0) {
                    builder.setAddress(socket.getInetAddress());
                } else {
                    InetAddress inetAddress = InetAddress.getByAddress(address);
                    if (inetAddress.equals(InetAddress.getByName("0.0.0.0"))) {
                        builder.setAddress(socket.getInetAddress());
                    } else {
                        builder.setAddress(inetAddress);
                    }
                }
                int zero = messageReader.getBuffer().getShort();
                checkArgument(zero == 0, "expected 0, found %s", zero);
                int port = messageReader.getBuffer().getShort();
                checkArgument(port > 0, "got invalid port value = %s", port);
                builder.setPort(port);
                int serverSocket = messageReader.getBuffer().getInt() & 1;
                builder.setServerSocket(serverSocket == 1);
                logger.debug("closing connection (temporary protocol mode)");
                return builder.build();
            }
        }
    }

    private static final class RelayDataOutputStream extends DataOutputStream {

        public RelayDataOutputStream(OutputStream out) {
            super(out);
        }

        private void writeHeader(int type, int length) throws IOException {
            writeInt(MAGIC);
            writeInt(type);
            writeInt(length);
        }

    }

    private static final class RelayDataInputStream extends DataInputStream {

        public RelayDataInputStream(InputStream in) {
            super(in);
        }

        public MessageReader readMessage() throws IOException {
            int magic = readInt();
            checkArgument(magic == MAGIC, "magic mismatch, got = %s, expected = %s", magic, MAGIC);
            int type = readInt();
            int length = readInt();
            checkArgument(length >= 0);
            ByteBuffer payload = ByteBuffer.allocate(length);
            IOUtils.readFully(this, payload.array());
            return new MessageReader(type, payload);
        }
    }

    private static final class Response {

        private final int code;
        private final String message;

        public Response(int code, String message) {
            this.code = code;
            this.message = message;
        }

        public int getCode() {
            return code;
        }

        public String getMessage() {
            return message;
        }

        @Override
        public String toString() {
            return "Response{" + "code=" + code + ", message=" + message + '}';
        }

    }

    private static final class MessageReader {

        private final int type;
        private final ByteBuffer buffer;

        public MessageReader(int type, ByteBuffer buffer) {
            this.type = type;
            this.buffer = buffer;
        }

        public int getType() {
            return type;
        }

        public ByteBuffer getBuffer() {
            return buffer;
        }

        public byte[] readLengthAndData() {
            int length = buffer.getInt();
            checkArgument(length >= 0);
            byte[] data = new byte[length];
            buffer.get(data);
            return data;
        }

        public Response readResponse() {
            int code = buffer.getInt();
            int messageLength = buffer.getInt();
            byte[] message = new byte[messageLength];
            buffer.get(message);
            return new Response(code, new String(message));
        }

        public MessageReader cloneReader() {
            return new MessageReader(type, ByteBuffer.wrap(buffer.array()));
        }

        private String dumpMessageForDebug() {
            if (type == RESPONSE) {
                return MoreObjects.toStringHelper("Response").add("code", cloneReader().readResponse().getCode()).add("message", cloneReader().readResponse().getMessage()).toString();
            } else {
                return MoreObjects.toStringHelper("Message").add("type", type).add("size", buffer.capacity()).toString();
            }
        }
    }

}

File: relay/src/main/kotlin/net/syncthing/java/client/protocol/rp/RelayClient.kt
/* 
 * Copyright (C) 2016 Davide Imbriaco
 *
 * This Java file is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.syncthing.java.client.protocol.rp

import com.google.common.base.MoreObjects
import com.google.common.base.Preconditions
import com.google.common.io.BaseEncoding
import net.syncthing.java.client.protocol.rp.beans.SessionInvitation
import net.syncthing.java.core.beans.DeviceAddress
import net.syncthing.java.core.beans.DeviceAddress.AddressType
import net.syncthing.java.core.configuration.ConfigurationService
import net.syncthing.java.core.interfaces.RelayConnection
import net.syncthing.java.core.security.KeystoreHandler
import org.apache.commons.io.IOUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory

import java.io.*
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket
import java.nio.ByteBuffer

import com.google.common.base.Preconditions.checkArgument
import net.syncthing.java.core.security.KeystoreHandler.*

class RelayClient(configuration: ConfigurationService) {
    private val logger = LoggerFactory.getLogger(javaClass)
    private val keystoreHandler: KeystoreHandler

    init {
        this.keystoreHandler = KeystoreHandler.Loader().loadAndStore(configuration)
    }

    @Throws(IOException::class, KeystoreHandler.CryptoException::class)
    fun openRelayConnection(address: DeviceAddress): RelayConnection {
        Preconditions.checkArgument(address.type == AddressType.RELAY)
        val sessionInvitation = getSessionInvitation(address.socketAddress, address.deviceId)
        return openConnectionSessionMode(sessionInvitation)
    }

    @Throws(IOException::class)
    fun openConnectionSessionMode(sessionInvitation: SessionInvitation): RelayConnection {
        logger.debug("connecting to relay = {}:{} (session mode)", sessionInvitation.address, sessionInvitation.port)
        val socket = Socket(sessionInvitation.address, sessionInvitation.port)
        val `in` = RelayDataInputStream(socket.getInputStream())
        val out = RelayDataOutputStream(socket.getOutputStream())
        run {
            logger.debug("sending join session request, session key = {}", sessionInvitation.key)
            val key = BaseEncoding.base16().decode(sessionInvitation.key)
            val lengthOfKey = key.size
            out.writeHeader(JOIN_SESSION_REQUEST, 4 + lengthOfKey)
            out.writeInt(lengthOfKey)
            out.write(key)
            out.flush()
        }
        run {
            logger.debug("reading relay response")
            val messageReader = `in`.readMessage()
            checkArgument(messageReader.type == RESPONSE)
            val response = messageReader.readResponse()
            logger.debug("response = {}", response)
            checkArgument(response.code == ResponseSuccess, "response code = %s (%s) expected %s", response.code, response.message, ResponseSuccess)
            logger.debug("relay connection ready")
        }
        return object : RelayConnection {
            override fun getSocket(): Socket {
                return socket
            }

            override fun isServerSocket(): Boolean {
                return sessionInvitation.isServerSocket
            }

        }
    }

    @Throws(IOException::class, KeystoreHandler.CryptoException::class)
    fun getSessionInvitation(relaySocketAddress: InetSocketAddress, deviceId: String): SessionInvitation {
        logger.debug("connecting to relay = {} (temporary protocol mode)", relaySocketAddress)
        keystoreHandler.createSocket(relaySocketAddress, KeystoreHandler.RELAY).use { socket ->
            RelayDataInputStream(socket.getInputStream()).use { `in` ->
                RelayDataOutputStream(socket.getOutputStream()).use { out ->
                    run {
                        logger.debug("sending connect request for device = {}", deviceId)
                        val deviceIdData = KeystoreHandler.deviceIdStringToHashData(deviceId)
                        val lengthOfId = deviceIdData.size
                        out.writeHeader(CONNECT_REQUEST, 4 + lengthOfId)
                        out.writeInt(lengthOfId)
                        out.write(deviceIdData)
                        out.flush()
                    }

                    run {
                        logger.debug("receiving session invitation")
                        val messageReader = `in`.readMessage()
                        logger.debug("received message = {}", messageReader.dumpMessageForDebug())
                        if (messageReader.type == RESPONSE) {
                            val response = messageReader.readResponse()
                            throw IOException(response.message)
                        }
                        checkArgument(messageReader.type == SESSION_INVITATION, "message type mismatch, expected %s, got %s", SESSION_INVITATION, messageReader.type)
                        val builder = SessionInvitation.Builder()
                                .setFrom(KeystoreHandler.hashDataToDeviceIdString(messageReader.readLengthAndData()))
                                .setKey(BaseEncoding.base16().encode(messageReader.readLengthAndData()))
                            val address = messageReader.readLengthAndData()
                        if (address.size == 0) {
                            builder.setAddress(socket.inetAddress)
                        } else {
                            val inetAddress = InetAddress.getByAddress(address)
                            if (inetAddress == InetAddress.getByName("0.0.0.0")) {
                                builder.setAddress(socket.inetAddress)
                            } else {
                                builder.setAddress(inetAddress)
                            }
                        }
                        val zero = messageReader.buffer.short.toInt()
                        checkArgument(zero == 0, "expected 0, found %s", zero)
                        val port = messageReader.buffer.short.toInt()
                        checkArgument(port > 0, "got invalid port value = %s", port)
                        builder.setPort(port)
                        val serverSocket = messageReader.buffer.int and 1
                        builder.setServerSocket(serverSocket == 1)
                        logger.debug("closing connection (temporary protocol mode)")
                        return builder.build()
                    }
                }
            }
        }
    }

    private class RelayDataOutputStream(out: OutputStream) : DataOutputStream(out) {

        @Throws(IOException::class)
        fun writeHeader(type: Int, length: Int) {
            writeInt(MAGIC)
            writeInt(type)
            writeInt(length)
        }

    }

    private class RelayDataInputStream(`in`: InputStream) : DataInputStream(`in`) {

        @Throws(IOException::class)
        fun readMessage(): MessageReader {
            val magic = readInt()
            checkArgument(magic == MAGIC, "magic mismatch, got = %s, expected = %s", magic, MAGIC)
            val type = readInt()
            val length = readInt()
            checkArgument(length >= 0)
            val payload = ByteBuffer.allocate(length)
            IOUtils.readFully(this, payload.array())
            return MessageReader(type, payload)
        }
    }

    private class Response(val code: Int, val message: String) {

        override fun toString(): String {
            return "Response{code=$code, message=$message}"
        }

    }

    private class MessageReader(val type: Int, val buffer: ByteBuffer) {

        fun readLengthAndData(): ByteArray {
            val length = buffer.int
            checkArgument(length >= 0)
            val data = ByteArray(length)
            buffer.get(data)
            return data
        }

        fun readResponse(): Response {
            val code = buffer.int
            val messageLength = buffer.int
            val message = ByteArray(messageLength)
            buffer.get(message)
            return Response(code, String(message))
        }

        fun cloneReader(): MessageReader {
            return MessageReader(type, ByteBuffer.wrap(buffer.array()))
        }

        fun dumpMessageForDebug(): String {
            return if (type == RESPONSE) {
                MoreObjects.toStringHelper("Response").add("code", cloneReader().readResponse().code).add("message", cloneReader().readResponse().message).toString()
            } else {
                MoreObjects.toStringHelper("Message").add("type", type).add("size", buffer.capacity()).toString()
            }
        }
    }

    companion object {

        private val MAGIC = -0x618643c0
        private val JOIN_SESSION_REQUEST = 3
        private val RESPONSE = 4
        private val CONNECT_REQUEST = 5
        private val SESSION_INVITATION = 6
        private val ResponseSuccess = 0
        private val ResponseNotFound = 1
        private val ResponseAlreadyConnected = 2
    }

}
